File: problems/LC/medium/python/15-3sum.py
# Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that
# i != j,
# i != k,
# j != k,
# nums[i] + nums[j] + nums[k] == 0.

# Notice that the solution set must not contain duplicate triplets.


# Example 1:
# Input: nums = [-1, 0, 1, 2, -1, -4]
# Output: [[-1, -1, 2], [-1, 0, 1]]

# Example 2:
# Input: nums = []
# Output: []

# Example 3:
# Input: nums = [0]
# Output: []


# Constraints:

# 0 <= nums.length <= 3000
# -105 <= nums[i] <= 105


from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
        iterate through list with left pointer
        use middle and right pointer to find if there's a complement
        """
        nums.sort()
        n = len(nums)
        ans = []
        for i in range(n - 1):
            # skips duplicate triplets
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            curr = nums[i]
            m, r = i + 1, len(nums) - 1

            while m < r:
                sum = nums[m] + nums[r] + curr

                if sum > 0:
                    r -= 1
                elif sum < 0:
                    m += 1
                else:
                    ans.append([curr, nums[m], nums[r]])
                    m += 1
                    r -= 1
                    while m < r and nums[m] == nums[m - 1]:
                        m += 1
        return ans

    def test(self, inputs, outputs):
        for i, input in enumerate(inputs):
            if (out := self.threeSum(*input)) != outputs[i]:
                print(f"{i} failed")
                print(f"Output:	 {out}")
                print(f"Correct: {outputs[i]}")
                print("")
                continue

            print(f"{i} pass")


input = [
    [[-1, 0, 1, 2, -1, -4]],
    [[0, 0, 0, 0, 0]],
    [[0]],
    [[-2, 0, 1, 1, 2]],
]
expected = [
    [[-1, -1, 2], [-1, 0, 1]],
    [[0, 0, 0]],
    [],
    [[-2, 0, 2], [-2, 1, 1]],
]


sol = Solution()
sol.test(input, expected)

File: problems/LC/medium/js/15-3-sum.ts
/**
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]]
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.

Example 1:
Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Explanation:
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].
Notice that the order of the output and the order of the triplets does not matter.

Example 2:
Input: nums = [0,1,1]
Output: []
Explanation: The only possible triplet does not sum up to 0.

Example 3:
Input: nums = [0,0,0]
Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.

Constraints:
3 <= nums.length <= 3000
-105 <= nums[i] <= 105
 */

const threeSum = (nums: number[]): number[][] => {
    const result: number[][] = [];

    nums.sort((a, b) => a - b);

    // fix 1 number, then run 2 sum
    for (let i = 0; i < nums.length - 1; i++) {
        // skips duplicate
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let l = i + 1,
            r = nums.length - 1,
            goal = nums[i];

        while (l < r) {
            const sum = nums[l] + nums[r];

            if (sum < -goal) {
                l += 1;
            } else if (sum > -goal) {
                r -= 1;
            } else {
                result.push([goal, nums[l], nums[r]]);
                l += 1;
                r -= 1;

                // skips duplicate
                while (nums[l] === nums[l - 1]) l++;
            }
        }
    }

    return result;
};

const test = () => {
    const input: [number[]][] = [
        [[-1, 0, 1, 2, -1, -4]],
        [[0, 1, 1]],
        [[0, 0, 0]],
    ];
    const output = [
        [
            [-1, -1, 2],
            [-1, 0, 1],
        ],
        [],
        [[0, 0, 0]],
    ];
    for (let i = 0; i < input.length; i++) {
        const result = threeSum(...input[i]);
        const correct = JSON.stringify(result) === JSON.stringify(output[i]);
        console.log(correct);
        if (!correct) {
            console.log("expected:", output[i]);
            console.log("actual:", result);
        }
        console.log();
    }
};

test();
export {};
